waiter/src/waiter/middleware.clj

```diff
@@ -0,0 +1,31 @@
+;;
+;;       Copyright (c) 2018 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.middleware
+  (:require [clojure.core.async :as async]
+            [full.async :refer (<?? <? go-try)]
+            [waiter.async-utils :as au]
+            [waiter.utils :as utils]))
+
+(defn wrap-context
+  "Wraps a handler, ensuring the request and response (or exception) contains data in the context map."
+  [handler context]
+  (fn [request]
+    (try
+      (let [response (handler (merge request context))]
+        (if (au/chan? response)
+          (async/go
+            (try
+              (merge (<? response) context)
```

Comment:
Why are we using `merge` instead of `assoc` with a specific key?

Reply:
Merge handles the case where I have a map or just a key/value pair.  That was my thought.  Is it worth an overload to replace
`(wrap-context {:descriptor descriptor})`
with
`(wrap-context :descriptor descriptor)`
I don't feel strongly that we need an overload.  

Additionally, I suspect this `merge` will eventually become a deep merge, in which case the signature doesn't change, but the `assoc`-style version is less useful.  We'd have to make an `assoc-in`-style version as well.